src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -44,1115 +44,724 @@
+        return false;
     }
 
-    return success;
-}
-
-
-////////////////////////////////////////////////////////////
-void JoystickImpl::close()
-{
-    if (directInput)
-        closeDInput();
+    vid = static_cast<USHORT>(vidLong);
+    pid = static_cast<USHORT>(pidLong);
+    return true;
 }
 
-////////////////////////////////////////////////////////////
-JoystickCaps JoystickImpl::getCapabilities() const
-{
-    if (directInput)
-        return getCapabilitiesDInput();
-
-    JoystickCaps caps;
-
-    caps.buttonCount = m_caps.wNumButtons;
-    if (caps.buttonCount > Joystick::ButtonCount)
-        caps.buttonCount = Joystick::ButtonCount;
-
-    caps.axes[Joystick::Axis::X]    = true;
-    caps.axes[Joystick::Axis::Y]    = true;
-    caps.axes[Joystick::Axis::Z]    = (m_caps.wCaps & JOYCAPS_HASZ) != 0;
-    caps.axes[Joystick::Axis::R]    = (m_caps.wCaps & JOYCAPS_HASR) != 0;
-    caps.axes[Joystick::Axis::U]    = (m_caps.wCaps & JOYCAPS_HASU) != 0;
-    caps.axes[Joystick::Axis::V]    = (m_caps.wCaps & JOYCAPS_HASV) != 0;
-    caps.axes[Joystick::Axis::PovX] = (m_caps.wCaps & JOYCAPS_HASPOV) != 0;
-    caps.axes[Joystick::Axis::PovY] = (m_caps.wCaps & JOYCAPS_HASPOV) != 0;
-
-    return caps;
-}
 
+} // namespace
 
-////////////////////////////////////////////////////////////
-Joystick::Identification JoystickImpl::getIdentification() const
+namespace sf::priv
 {
-    return m_identification;
-}
-
-
 ////////////////////////////////////////////////////////////
-JoystickState JoystickImpl::update()
+void JoystickImpl::dispatchDeviceConnected(HANDLE deviceHandle)
 {
-    if (directInput)
+    JoystickImpl joystickImpl{};
+    joystickImpl.m_lastDeviceHandle = deviceHandle;
+    joystickImpl.m_index            = static_cast<std::uint32_t>(joysticks.size());
+    joystickImpl.m_state.connected  = true;
+
+    UINT nameSize{};
+    // okay this looks weird, but the docs say to call it twice like this
+    GetRawInputDeviceInfoW(deviceHandle, RIDI_DEVICENAME, nullptr, &nameSize);
+    GetRawInputDeviceInfoW(deviceHandle, RIDI_DEVICENAME, deviceNameDataChunk.data(), &nameSize);
+    std::wstring devicePath(reinterpret_cast<LPCWSTR>(deviceNameDataChunk.data()));
+    auto*        fileHandle = CreateFileW(devicePath.data(),
+                                   GENERIC_READ | GENERIC_WRITE,
+                                   FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                   nullptr,
+                                   OPEN_EXISTING,
+                                   0x0,
+                                   nullptr);
+    if (HidD_GetProductString(fileHandle, deviceHumanNameDataChunk.data(), rawInputChunkSize))
     {
-        if (m_buffered)
+        joystickImpl.m_identification.name = String{reinterpret_cast<wchar_t*>(deviceHumanNameDataChunk.data())};
+        USHORT vid{};
+        USHORT pid{};
+        bool   isXInput = false;
+        if (extractVidPid(devicePath, vid, pid, isXInput))
```

Comment:
Perhaps instead of this function returning a bool + using output parameter, we could put all the output values into a structure then return a `std::optional<SomeNewStruct>`. Then we can do something like this
```cpp
if (const auto foo = extractVidPid(devicePath))
{
    joystickImpl.m_identification.vendorId  = foo->vid;
    joystickImpl.m_identification.productId = foo->pid;
    joystickImpl.m_useXInput                = true;
}
```

Reply:
I went ahead and made this change but oddly `is`XInput` is never used. Is that a bug or should we remove that variable?